rfcs/rfc-colin-states-mapping-to-lear.md

```diff
@@ -0,0 +1,127 @@
+}
+```
+
+An invalid business state transition map will be implemented as a python dictionary or database table.  This map can be used in conjunction with the existing [ALLOWABLE_FILINGS](https://github.com/bcgov/lear/blob/e850f9a22672910db6e4ceb2f1ddb9437541a86f/legal-api/src/legal_api/services/authz.py#L96) dictionary as well as any other conditions to determine what filings are allowed for a business at a given point in time.
+
+_Example of what invalid state transition might look like_
+``` python
+INVALID_STATE_TRANSITION_FILINGS: Final = {
+    'staff': {
+        Business.State.ACTIVE: { # current business state
+            'restoration': { # target filing
+                # invalid target filing if filing that transitioned business to its current state is in list
+                'fullRestoration': ['registration', 'incorporationApplication'], 
+                'limitedRestoration': ['registration', 'incorporationApplication'], 
+            }
+            ...
+        },
+        Business.State.HISTORICAL: {
+            'restoration': { 
+                'fullRestoration': ['almagmation', 'continuationOut', 'continuationIn'], 
+                'limitedRestoration': ['almagmation', 'continuationOut', 'continuationIn'],
+            },
+            'putBackOn': ['almagmation', 'withdrawal']
+            ...
+        },
+        Business.State.LIQUIDATION: {
+            'restoration': {
+                'fullRestoration': ['liquidation'], 
+                'limitedRestoration': ['liquidation'], 
+            },
+            ...
+        }
+    },
+    'client': {
+        Business.State.ACTIVE: { # current business state
+            ...
+        },
+        Business.State.HISTORICAL: {
+            ...
+        },
+        Business.State.LIQUIDATION: {
+            'restoration': {
+                'fullRestoration': ['liquidation'], 
+                'limitedRestoration': ['liquidation'], 
+            },
+            'putBackOn': ['liquidation']
+            ...
+        }
+    }        
+    
+}
+```
+
+_Allowable state transitions matrix in COLIN for restoration and put back on filings_
+![colin_allowable_filings.png](rfc-colin-states-mapping-to-lear/colin_allowable_filings.png)
+
+Filing validation logic will be updated to include logic to ensure that only valid business state transitions are allowed. 
+
+All filings resulting in business state transitions will need to populate the businesses.state_filing_id field.  This is used to determine if future business state transitions are permitted.
+
+Filings.meta_data field for any filings where a business state transitions occurs will include from/to state.
+
+API consumers such as the FE with a need to determine which filing transitioned the business to its current state will retrieve the business state filing provided by the business response of the GET business endpoint.  For example, this can be used to determine if an ACTIVE business is currently in _limited restoration_.
```

Comment:
To determined if a business is currently in Limited Restoration, could we not simply check if State=LRS ?

Reply:
LRS is not a business state.  Currently, the only business states we have are ACTIVE, HISTORICAL and LIQUIDATION.

Reply:
OK, so whenever there is a State Filing ID in the business object then the FE will need to fetch that filing to determine what the current business "state" is?

I'm wondering if it would be better for the API to provide another state variable with this info so that the UI doesn't have to make another network call? I know this going back to the COLIN model but it's more work for the UI and the API already has that data available to it. Thoughts?

Reply:
Here is what the UI does if the business state is Historical. This will have to be updated/enhanced. (The Reason Text that is set on line 562 is displayed in the Entity Info header.)
https://github.com/bcgov/business-filings-ui/blob/697c7c44365157113462cb95b708fa626e2e8703/src/App.vue#L534

If state filing summary info (or general business summary info) was provided then the extra call to fetch the state filing would not be needed -- in this scenario. It may still be needed in other scenarios.

How about listing out a few examples of when the FE needs state filing info?

Reply:
Actually, I think we can just return what the `/filings` endpoint returns for the state filing.  For some reason I thought it was a larger payload.  Then you can use `parseStateFiling` function on the FE as is without making an extra call.

Reply:
Yes, that would save the extra call. (There will be a development dependency: the API will have to be updated with the new object first and then the UI can switch over to it.)